### PR TITLE
make string_of_float and float_of_string locale-independent

### PR DESCRIPTION
**Pull request by ygrek**
Problem:

see https://caml.inria.fr/mantis/view.php?id=6701 and https://github.com/mjambon/yojson/issues/13

Solution:

1. Use `uselocale` around calls to `vsnprintf`.
2. Use `strtod_l` if available. If not available - use `uselocale` around calls to `strtod`.

Possible improvement - use `vsnprintf_l` if available.

Test output for 4.04.1 : 
```
locale from environment
  print 1.23 : 1.23
  parse "12345.6789" : 12345.6789
  roundtrip 1.23 : 1.23
locale nl_NL
  print 1.23 : 1,23
  parse "12345.6789" : exn Failure("float_of_string")
  roundtrip 1.23 : 1,23
locale POSIX
  print 1.23 : 1.23
  parse "12345.6789" : 12345.6789
  roundtrip 1.23 : 1.23
```

With this patch : 
```
$ ./q.locale 
locale from environment
  print 1.23 : 1.23
  parse "12345.6789" : 12345.6789
  roundtrip 1.23 : 1.23
locale nl_NL
  print 1.23 : 1.23
  parse "12345.6789" : 12345.6789
  roundtrip 1.23 : 1.23
locale POSIX
  print 1.23 : 1.23
  parse "12345.6789" : 12345.6789
  roundtrip 1.23 : 1.23
```

Speed overhead of `uselocale` around `vsnprintf` (x86_64, glibc).
Test program : 
```
let () =
  for _ = 1 to 1_000_000 do
    ignore @@ Printf.sprintf "%f %f" 123456.789 (-8888.8888)
  done
```

```
$ sudo perf stat ./speed.4.04.1 |& grep instructions
     9,684,761,712      instructions              #    1.72  insn per cycle         
$ sudo perf stat ./speed.locale |& grep instructions
     9,780,738,205      instructions              #    1.73  insn per cycle         
```

i.e. each call to `vsnprintf` (~4800 instructions) bears additional cost of ~48 insns for two `uselocale` calls, i.e. 1% slowdown for every `%f` conversion in `Printf.printf`.

### CODE REVIEW

**Comment by gasche** [comment]
The code looks sensible to me, and the benchmark is reassuring (I wondered about the performance impact), but the Windows CI seems to be failing.

File: byterun/floats.c
Comment (damiendoligez) [inline byterun/floats.c]: typo: "third-party"

File: byterun/floats.c
Comment (damiendoligez) [inline byterun/floats.c]: Isn't `locale_t` a pointer type? I'd much rather make that explicit and use `NULL` instead of `0` (here and in a few other places).

File: byterun/floats.c
Comment (damiendoligez) [inline byterun/floats.c]: This will leak memory on OSX (and presumably the BSDs), whose `duplocale`/`newlocale` API is much more reasonable than Linux's.

**Review by damiendoligez — DISMISSED** [review]
Reviewed and made a few comments.

File: byterun/floats.c
Comment (ygrek) [reply byterun/floats.c]: Good catch!
I didn't know these api is such a mess. Funny thing is BSD claims it conforms to POSIX 2008 too, go figure :]
I believe I can get away with dropping duplocale usage and doing `newlocale(LC_NUMERIC_MASK,"C",(locale_t)0)`

File: byterun/floats.c
Comment (ygrek) [reply byterun/floats.c]: man page uses `(locale_t)0` - I will use that too

**Comment by ygrek** [comment]
AFAIU mingw port links against old msvcrt which doesn't have _create_locale, but thankfully has per-thread setlocale

**Comment by ygrek** [comment]
ftr my work here is done at this point, waiting for review/decision

**Comment by xavierleroy** [comment]
I think we're at a fork in the road: either we go the way of this PR, with all the nonsensical APIs offered by various C standard libraries, or we find and reuse good code that performs float<->string conversions the way we want them.  I'm still punting on this decision.  Opinions welcome.

**Comment by dbuenzli** [comment]
It would generally make it easier to embed the runtime system in exotic settings (e.g. for unikernels), the hard libc requirement of the ocaml runtime system are very minimal (`string.h`, `snprintf` and `malloc`, the rest can be stubbed by returning `ENOSYS`) and that's one of the painful one to provide (I and others have been using [this code](http://www.netlib.org/fp/dtoa.c) for that).

There was a bit of discussion in [MPR7218](https://caml.inria.fr/mantis/view.php?id=7218) with links to recent floating point conversion routines implementations. Also following the related issues shows that floating point conversion issue seem to show up quite recurrently.

**Comment by avsm** [comment]
I'd like to second the desire to not depend on the OS runtime for this if at all possible. It's been responsible for a surprising amount of the complexity of embedding OCaml in non-libc environments. The `dtoa` implementation that @dbuenzli linked above has worked pretty well, and is hopefully more readable when unifdef'ed.

**Comment by gasche** [comment]
I like the gained consistency/portability, but `string_of_float` can be a performance-sensitive operation for many script-like data-moving programs so one should also ensure that there is no notable performance regression compared to the libc we care about.

**Comment by dbuenzli** [comment]
Regarding performance it seems both the V8 and SpiderMonkey JavaScript engines replaced Gay's code linked above with [this code](https://github.com/google/double-conversion) for improving performance, given the setting in which these operate it should also give a good confidence w.r.t. correctness, but the problem is that it's written in C++. Someone [here](https://github.com/miloyip/dtoa-benchmark) seems to have benchmarked a few dtoa implementations.

**Comment by ygrek** [comment]
FTR I would also prefer embedding code given that system api is so braindead and inconsistent (but I will not be working on it - too much work)

**Comment by alainfrisch** [comment]
I would also like to embed battle-tested implementation of string<->float conversion functions, but this won't be completely straightforward (need to choose which implementation, discuss if the license is ok, benchmark, etc) and nobody has volunteered to drive this project.  So I'm in favor of already accepting this PR, which fixes an actual problem and has already been reviewed. 
 @damiendoligez : do you agree?

**Comment by damiendoligez** [comment]
@alainfrisch You're right. Let's take this branch of the fork and we can later backtrack and embed the functions when we find good implementations with an acceptable license.

**Review by damiendoligez — APPROVED** [review]
Still looks good.

**Comment by damiendoligez** [comment]
Formally approved. I'll merge in a few days if there are no objections.

**Comment by garrigue** [comment]
On MacOs, I get errors and cannot compile `float.c`.
The compiler says that `locale_t` and `use locale` are undefined.

**Comment by objmagic** [comment]
+1 on Jacque's comment. 

Just merged trunk and `float.c` cannot compile now. I am on Mac OS X 10.13.3.

**Comment by gasche** [comment]
I just reverted the PR ( revert commit 682aceb2a470ee314170f79c63b27639ce5c1983 ) so that everyone can keep using trunk. I will re-submit a PR so that we don't forget that it has to be re-merged.

**Comment by gasche** [comment]
@garrigue, @objmagic: just to be sure, you did re-run `./configure` before trying to build from the new `trunk`?

**Comment by objmagic** [comment]
@gasche Yes. `make distclean` -> `./configure` -> `make world.opt`

**Comment by shindere** [comment]
Jacques Garrigue (2018/03/28 19:14 -0700):
> On MacOs, I get errors and cannot compile `float.c`.
> The compiler says that `locale_t` and `use locale` are undefined.

Yeah Inria's CI is broken, too, for the same reason.

**Comment by shindere** [comment]
It would be good to let the new PR go through precheck before merging.

**Comment by shindere** [comment]
Gabriel Scherer (2018/03/29 05:08 +0000):
> @garrigue, @objmagic: just to be sure, you did re-run `./configure`
> before trying to build from the new `trunk`?

At least CI does so

**Comment by gasche** [comment]
This is now re-merged as #1803.

**Pull request merged**
